internal/service/project/model_project.go

```diff
@@ -202,3 +206,38 @@ func UpdateProjectBool(plan, state types.Bool, setting **bool) bool {
 	}
 	return false
 }
+
+func NewTFTags(tags []admin.ResourceTag, useNullForEmpty bool) types.Map {
+	if tags == nil {
+		if useNullForEmpty {
```

Comment:
[q] Why are we using `useNullForEmpty`?

Reply:
It took me a while to wrap my head around this 😅

We need the flag `useNullForEmpty` to separate two cases:

1. `tags` is not specified:
```hcl
resource "mongodbatlas_project" "tags_example" {
  name   = "tags-readme-example1"
  org_id = var.org_id
}
```

2. `tags` is specified but empty

```hcl
resource "mongodbatlas_project" "tags_example" {
  name   = "tags-readme-example1"
  org_id = var.org_id
  tags = {}
}
```

When calling `project.GetTags()` in the SDK, we **always** get an empty `[]admin.ResourceTag{}.`
To avoid provider bugs such as:

![image](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/65eb5559-9616-45a3-a0ce-36ba9afc0384)

Does that answer your question?

Reply:
This is related to the removal of `computed`, see [discussion above](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2135#discussion_r1560681693)

After a discussion with @lantoli and @maastha offline, I have the following finding:

> Using `Optional` and `Computed` simplifies the code and the user experience:

1. User experience: Cases 1 & 2 above lead to the same Terraform state. By introducing `tags = {}` there will be no state changes.
2. Code simplified: We don't have to differentiate between Null and EmptyList. We can then use `ProjectAPi.GetTags()`, which always returns an empty list and never nil.

WDYT? @AgustinBettati @maastha @marcosuma 
Can I go back to using both `Optional` and `Computed`?

Reply:
So if we only use `project.Tags` instead of `project.GetTags()` and have Tags attribute as optional-only, can the issue with empty `tags = {}` be handled better? would we still need `useNullForEmpty`?

Reply:
One scenario where using Optional & computed might be problematic is if user wants to delete all created tags, then if they remove the tags attribute completely from their config to do this, TF might not detect a plan diff as it could interpret the previous value in the state as a "computed" value, @EspenAlbert can you confirm if this happens?

Reply:
my understanding is the API response doesn't  return an empty tag list, but the GetTags() method does

Reply:
### Understanding the impact of `computed`

#### Understanding what is the api response

1. What happens when we create a project with no tags
2. What happens when we create a project with empty tags

- In both cases, when reading the project (either by name or id), tags are always set to the empty list (`"tags": []` in the JSON response)

If changes happen outside terraform, will terraform plan detect the change?

#### Case 1: Create a project, update a new tag from the UI. Result: Changes will be detected in the terraform plan and reverted after terraform apply

```hcl
resource "mongodbatlas_project" "tags_test2" {
  name   = "espen-tags-test"
  org_id = var.org_id

  tags = {
    Owner       = "Terraform"
    Environment = "Example"
  }
}
```
![image-1](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/e70a2b2b-df58-45d4-bfc4-42e04753a31c)

- Updating the value of `Environment=UI` and adding a `UI=yes` tag

![image-2](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/c032288d-e3e6-41d7-9a4d-b186490cb86e)

- re-running `terraform apply`

![image-3](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/df6bf21e-3c8d-463a-b9bb-b56f1efd78d2)

- UI after `yes`
  
![image-4](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/ef6bce28-e94f-4488-a31e-44c739ae31db)


#### Case 2: Import a project and run without the `tags` attribute. Result: Changes will be detected and terraform will remove the tags

- Existing project with tags

![image-4](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/bf65a295-f60f-47d0-9ad2-25ad448cbded)

- terraform configuration

```hcl
resource "mongodbatlas_project" "tags_test2" {
  name   = "espen-tags-test"
  org_id = var.org_id
}
```

- Import the project

![image-6](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/f9515068-df3f-4a5c-b22d-4d97e4caaf48)

- running `terraform apply`

![image-5](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/85b77a82-d8e3-4722-a1e6-aeeb9b167459)

- UI after `yes`

![image-7](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/cf5adae3-c01e-43b6-869b-9edfae0301d6)

Reply:
I will research how AWS handles the tags, but for now, I think there are no concerns about having them `Computed` since Terraform will always use the Terraform configuration file as the source of truth

Reply:
I am struggling to understand why in `case 2` a plan diff is shown when the attribute is marked as computed and not defined in the config.
For reference was able to find in [google best practices](https://github.com/hashicorp/terraform-provider-google/wiki/Developer-Best-Practices#optionalcomputed-lists-of-objects) a mention related to this risk (`removing an attribute completely from the config and having no plan diff`) when having a list of objects that are optional + computed. The guide mentions this usage for cases when there are default values coming from the API, but I do not believe this would be the case.

Reply:
>I am struggling to understand why in case 2 a plan diff is shown when the attribute is marked as computed

@AgustinBettati I think the plan diff is shown because for Computed value it is going to show` (known after apply)`, notice that same happens for ip_addresses. The actual diff seems to be triggered by `with_default_alert_settings` which has a default value i think.


Also, based on the cases that we are looking at here, it seems we see adding `Computed` because we want to understand if it can potentially handle changes outside of Terraform (is that correct @EspenAlbert ?) If yes, then we don't need to that because that would imply every attribute in every resource should always be marked as Computed.

@EspenAlbert  I think we can close on this over a call with the team WDYT?

Reply:
@AgustinBettati 
Tried these two cases here:

### Case 3: Create a project without tags, add a tag in the UI, result: `no changes`

However, if we change the project name, then the tags will be updated to empty

### Case 4: Create project without tags, add a tag in the UI, update tags to `tags = {}`, result: Changes seen and removed

Reply:
@maastha no, I was trying to address the `cons` of using `Computed` with the cases above. 
Right now, I cannot see the `cons` of using `Computed`.

The "pros" for adding `Computed` are (also mentioned [in the comment above](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2135#discussion_r1562698554)):

1. Improve user experience: Avoid plan change when `tags` is not defined  -> and `tags = {}`
2. Improve the code readability. Can always represent tags as an empty list/map in the state. (Simplifies code since the API always returns `tags=[]` when tags are not defined or empty)

Looking forward to the call 😁

Reply:
Discussed offline and resolved ✅